README.md
# MTE 544 Assignment

Kalman filter assignment questions simulated in Haskell. Uses the [Linear](https://hackage.haskell.org/package/linear) for linear algebra operations and [EasyPlot](https://hackage.haskell.org/package/easyplot) for plotting.

The build tool used is [Stack](https://docs.haskellstack.org/en/stable/README/). Build and run using the following:

```
stack build
stack run mte544-exe
```

Resulting plots can be seen in the `image` directory.


ChangeLog.md
# Changelog for mte544

## Unreleased changes


src/Q3.hs
module Q3
    ( leftEncoder
    , rightEncoder
    , getControl
    , initialState
    , toTup
    , repeatStep
    , getRMSE
    ) where

import Control.Lens.Getter ((^.))

import           Linear

type State = V3 Float
type Control = V2 Float

dt :: Float
dt = 0.10

time :: [Float]
time = iterate (+ dt) 0

leftEncoder :: [Float]
leftEncoder = map (cos . (/2)) time

rightEncoder :: [Float]
rightEncoder = map (sin . (/2)) time

dCenter :: Float -> Float -> Float
dCenter l r = (l + r) / 2

dTheta :: Float -> Float -> Float
dTheta l r = (r - l) / 0.45

diffSteps :: [Float] -> [Float]
diffSteps = zipWith (-) =<< tail

getControl :: [Float] -> [Float] -> [Control]
getControl ls rs = zipWith (\a b -> V2 (dCenter a b) (dTheta a b)) (diffSteps ls) (diffSteps rs)

initialState :: State
initialState = V3 0.1 0.2 0.7854

updateStateX :: State -> Control -> State
updateStateX x u = x + V3 (u ^._x * cos (x ^._z + u ^._y / 2))
                          (u ^._x * sin (x ^._z + u ^._y / 2))
                          (u ^._y)

toTup :: State -> (Float, Float)
toTup sp = (sp ^._x, sp ^._y)

repeatStep :: Int -> State -> [Control] -> [State]
repeatStep n' sp' inp' = repeatStep' n' sp' inp' 0

repeatStep' :: Int -> State -> [Control] -> Int -> [State]
repeatStep' n sp inp i
  | i == n = []
  | otherwise = sp' : repeatStep' n sp' us (i+1)
  where (u : us) = inp
        sp' = updateStateX sp u

getDistance :: State -> State -> Float
getDistance x1 x2 = euclidean (x1 ^._x - x2 ^._x) (x1 ^._y - x2 ^._y)
  where euclidean a b = sqrt (a * a + b * b)

getRMSE :: [State] -> [State] -> [(Float, Float)]
getRMSE sp1 sp2 = zip time (zipWith getDistance sp1 sp2)


src/Q4a.hs
module Q4a
    ( initialState
    , controlInput
    , repeatKFStep
    , toStateX
    , printState
    ) where

import Control.Lens.Getter ((^.))
import Data.Random.Normal

import           Linear
import           Linear.Matrix

type State = V3 Float
type Control = V2 Float
type Covariance = M33 Float
type Measurement = V3 Float

data StatePose = StatePose { stateX :: State
                           , stateP :: Covariance } deriving (Show, Eq)

randMeasure :: [Float]
randMeasure = mkNormals 11111

dt :: Float
dt = 0.10

time :: [Float]
time = iterate (+ dt) dt

controlInput :: [Control]
controlInput = map (\x -> V2 (1 + (sin x)) (sin x)) time

initialState :: StatePose
initialState = StatePose (V3 0 0 0) (identity !!* 0.01 :: Covariance)

rmat :: Covariance
rmat = identity

qmat :: Covariance
qmat = V3 (V3 0.5 0 0)
          (V3 0 0.5 0)
          (V3 0 0 1 :: V3 Float)

createMeasNoise :: Int -> Float -> Measurement
createMeasNoise i q = (sqrt <$> qt) !* randVec
  where randVec = V3 (randMeasure !! (i * 3)) (randMeasure !! (i * 3 + 1))
                     (randMeasure !! (i * 3 + 2))
        qt = qmat !!* q

updateStateX :: Control -> State -> State
updateStateX u x = x + V3 (dt * u ^._x * cos (x ^._z))
                          (dt * u ^._x * sin (x ^._z))
                          (dt * u ^._y)

getStateTransition ::  Control -> State -> Covariance
getStateTransition u x = V3 (V3 1 0 (-dt * u ^._x * sin (x ^._z)))
                            (V3 0 1 (dt * u ^._x * cos (x ^._z)))
                            (V3 0 0 1 :: V3 Float)

hFunc :: State -> Measurement
hFunc = id

hJacobian :: State -> M33 Float
hJacobian _ = identity

predict :: Control -> StatePose -> Float -> StatePose
predict u sp r = StatePose (updateStateX u (stateX sp)) ((f !*! p !*! (transpose f)) !+! rt)
  where p = stateP sp
        rt = rmat !!* r
        f = getStateTransition u (stateX sp)

update :: Measurement -> StatePose -> Float -> StatePose
update z sp q = StatePose (x + k !* y) (((identity :: M33 Float) !-! (k !*! h)) !*! p)
  where x = stateX sp           :: State
        p = stateP sp           :: Covariance
        y = z - (hFunc x)       :: Measurement
        h = hJacobian x         :: M33 Float
        b = (h !*! p !*! (transpose h)) !+! (qmat !!* q) :: M33 Float
        k = p !*! (transpose h) !*! (inv33 b) :: M33 Float

fakeMeasurements :: State -> Measurement -> Measurement
fakeMeasurements x noise = hFunc x + noise

kfStep :: Control -> Measurement -> StatePose -> Float -> Float -> StatePose
kfStep u mNoise sp r q = update z sp' q
  where sp' = predict u sp r
        z = fakeMeasurements (stateX sp') mNoise

toStateX :: StatePose -> (Float, Float)
toStateX sp = (stateX sp ^._x, stateX sp ^._y)

printState :: StatePose -> IO ()
printState sp = putStrLn ("X: " ++ show x ++ " Y: " ++ show y)
  where x = (stateX sp) ^._x
        y = (stateX sp) ^._y

repeatKFStep :: Int -> StatePose -> [Control] -> Float -> Float -> [StatePose]
repeatKFStep n' sp' inp' r q = repeatKFStep' n' sp' inp' r q 0

repeatKFStep' :: Int -> StatePose -> [Control] -> Float -> Float -> Int -> [StatePose]
repeatKFStep' n sp inp r q i
  | i == n = []
  | otherwise = sp' : repeatKFStep' n sp' fs r q (i+1)
  where (f : fs) = inp
        mNoise = createMeasNoise i q
        sp' = kfStep f mNoise sp r q


src/Q4b.hs
module Q4b
    ( initialState
    , controlInput
    , repeatKFStep
    , toStateX
    , printState
    ) where

import Control.Lens.Getter ((^.))
import Data.Random.Normal

import           Linear
import           Linear.Matrix

type State = V3 Float
type Control = V2 Float
type Covariance = M33 Float
type Measurement = V2 Float

data StatePose = StatePose { stateX :: State
                           , stateP :: Covariance } deriving (Show, Eq)

randMeasure :: [Float]
randMeasure = mkNormals 11111

dt :: Float
dt = 0.10

time :: [Float]
time = iterate (+ dt) dt

controlInput :: [Control]
controlInput = map (\x -> V2 (1 + (sin x)) (sin x)) time

initialState :: StatePose
initialState = StatePose (V3 0 0 0) (identity !!* 0.01 :: Covariance)

rmat :: Covariance
rmat = identity

qmat :: M22 Float
qmat = V2 (V2 0.5 0)
          (V2 0 1 :: V2 Float)

createMeasNoise :: Int -> Float -> Measurement
createMeasNoise i q = (sqrt <$> qt) !* randVec
  where randVec = V2 (randMeasure !! (i * 2)) (randMeasure !! (i * 2 + 1))
        qt = qmat !!* q

updateStateX :: Control -> State -> State
updateStateX u x = x + V3 (dt * u ^._x * cos (x ^._z))
                          (dt * u ^._x * sin (x ^._z))
                          (dt * u ^._y)

getStateTransition ::  Control -> State -> Covariance
getStateTransition u x = V3 (V3 1 0 (-dt * u ^._x * sin (x ^._z)))
                            (V3 0 1 (dt * u ^._x * cos (x ^._z)))
                            (V3 0 0 1 :: V3 Float)

hFunc :: State -> Measurement
hFunc x = V2 (sqrt $ (x ^._x) ^ 2 + (x ^._y) ^ 2) (tan (x ^._z))

hJacobian :: State -> M23 Float
hJacobian xVec = V2 (V3 (x / sqrtxy) (y / sqrtxy) 0)
                    (V3 0 0 (tanth ^ 2 + 1))
  where x = xVec ^._x
        y = xVec ^._y
        tanth = tan $ xVec ^._z
        sqrtxy = sqrt $ x ^ 2 + y ^ 2

predict :: Control -> StatePose -> Float -> StatePose
predict u sp r = StatePose (updateStateX u (stateX sp)) ((f !*! p !*! (transpose f)) !+! rt)
  where p = stateP sp
        rt = rmat !!* r
        f = getStateTransition u (stateX sp)

update :: Measurement -> StatePose -> Float -> StatePose
update z sp q = StatePose (x + k !* y) (((identity :: M33 Float) !-! (k !*! h)) !*! p)
  where x = stateX sp           :: State
        p = stateP sp           :: Covariance
        y = z - (hFunc x)       :: Measurement
        h = hJacobian x         :: M23 Float
        b = (h !*! p !*! (transpose h)) !+! (qmat !!* q) :: M22 Float
        k = p !*! (transpose h) !*! (inv22 b) :: M32 Float

fakeMeasurements :: State -> Measurement -> Measurement
fakeMeasurements x noise = hFunc x + noise

kfStep :: Control -> Measurement -> StatePose -> Float -> Float -> StatePose
kfStep u mNoise sp r q = update z sp' q
  where sp' = predict u sp r
        z = fakeMeasurements (stateX sp') mNoise

toStateX :: StatePose -> (Float, Float)
toStateX sp = (stateX sp ^._x, stateX sp ^._y)

printState :: StatePose -> IO ()
printState sp = putStrLn ("X: " ++ show x ++ " Y: " ++ show y)
  where x = (stateX sp) ^._x
        y = (stateX sp) ^._y

repeatKFStep :: Int -> StatePose -> [Control] -> Float -> Float -> [StatePose]
repeatKFStep n' sp' inp' r q = repeatKFStep' n' sp' inp' r q 0

repeatKFStep' :: Int -> StatePose -> [Control] -> Float -> Float -> Int -> [StatePose]
repeatKFStep' n sp inp r q i
  | i == n = []
  | otherwise = sp' : repeatKFStep' n sp' fs r q (i+1)
  where (f : fs) = inp
        mNoise = createMeasNoise i q
        sp' = kfStep f mNoise sp r q


app/Main.hs
module Main where

import           Graphics.EasyPlot
import qualified Q2
import qualified Q3
import qualified Q4a
import qualified Q4b

-- Problem 2 with Varying R
q2r :: IO Bool
q2r = plot (PNG "image/q2r.png") $ map (\(x, y) -> Data2D [Title ("R = " ++ show x), Color y, Style Lines] [] (map Q2.toStateX (states x 1.0))) [(0, Red), (0.001, Blue), (0.01, Green)]
  where states = Q2.repeatKFStep 100 Q2.initialState Q2.controlInput

-- Problem 2 with Varying Q
q2q :: IO Bool
q2q = plot (PNG "image/q2q.png") $
  map (\(x, y) -> Data2D [Title ("Q = " ++ show x ++ " * Qt"),
                          Color y,
                          Style Lines] [] (map Q2.toStateX (states 0 x)))
      [(0.5, Red), (1, Blue), (2, Green)]
  where states = Q2.repeatKFStep 100 Q2.initialState Q2.controlInput

-- Problem 3 trajectory and rmse plots
q3 :: IO Bool
q3 = do
  _ <- plot (PNG "image/q3_traj.png")
    [ Data2D [Title "Original", Color Red, Style Lines] [] (map Q3.toTup cleanStates)
    , Data2D [Title "2%/2%", Color Blue, Style Lines] [] (map Q3.toTup noisyState1)
    , Data2D [Title "3%/2%", Color Green, Style Lines] [] (map Q3.toTup noisyState2)]
  _ <- plot (PNG "image/q3_rmse.png")
    [ Data2D [Title "2%/2%", Color Red, Style Points] [] (Q3.getRMSE cleanStates noisyState1)
    , Data2D [Title "3%/2%", Color Blue, Style Points] [] (Q3.getRMSE cleanStates noisyState2)]
  return True
  where cleanControl = Q3.getControl Q3.leftEncoder Q3.rightEncoder
        noisyControl1 = Q3.getControl (map (* 1.02) Q3.leftEncoder) (map (* 1.02) Q3.rightEncoder)
        noisyControl2 = Q3.getControl (map (* 1.03) Q3.leftEncoder) (map (* 1.02) Q3.rightEncoder)
        cleanStates = Q3.repeatStep 100 Q3.initialState cleanControl
        noisyState1 = Q3.repeatStep 100 Q3.initialState noisyControl1
        noisyState2 = Q3.repeatStep 100 Q3.initialState noisyControl2

-- Problem 4 with measurement A varying R
q4ar :: IO Bool
q4ar = plot (PNG "image/q4ar.png") $ map (\(x, y) -> Data2D [Title ("R = " ++ show x), Color y, Style Lines] [] (map Q4a.toStateX (states x 1))) [(0, Red), (0.001, Blue), (0.01, Green)]
  where states = Q4a.repeatKFStep 200 Q4a.initialState Q4a.controlInput

-- Problem 4 with measurement B varying R
q4br :: IO Bool
q4br = plot (PNG "image/q4br.png") $ map (\(x, y) -> Data2D [Title ("R = " ++ show x), Color y, Style Lines] [] (map Q4b.toStateX (states x 1))) [(0, Red), (0.001, Blue), (0.01, Green)]
  where states = Q4b.repeatKFStep 200 Q4b.initialState Q4b.controlInput

-- Problem 4 with measurement A varying Q
q4aq :: IO Bool
q4aq = plot (PNG "image/q4aq.png") $ map (\(x, y) -> Data2D [Title ("Q = " ++ show x ++ " * Qt"), Color y, Style Lines] [] (map Q4a.toStateX (states 0 x))) [(0.5, Red), (1, Blue), (2, Green)]
  where states = Q4a.repeatKFStep 200 Q4a.initialState Q4a.controlInput

-- Problem 4 with measurement B varying Q
q4bq :: IO Bool
q4bq = plot (PNG "image/q4bq.png") $ map (\(x, y) -> Data2D [Title ("Q = " ++ show x ++ " * Qt"), Color y, Style Lines] [] (map Q4b.toStateX (states 0 x))) [(0.5, Red), (1, Blue), (2, Green)]
  where states = Q4b.repeatKFStep 200 Q4b.initialState Q4b.controlInput

main :: IO ()
main = sequence_ $ [q2r, q2q, q3, q4ar, q4br, q4aq, q4bq]



src/Q2.hs
module Q2
    ( initialState
    , controlInput
    , repeatKFStep
    , toStateX
    , printState
    ) where

import Control.Lens.Getter ((^.))
import Data.Random.Normal

import           Linear
import           Linear.Matrix

type State = V4 Float
type Control = V2 Float
type Covariance = M44 Float
type Measurement = V2 Float

data StatePose = StatePose { stateX :: State
                           , stateP :: Covariance } deriving (Show, Eq)
randMeasure :: [Float]
randMeasure = mkNormals 11111

dt :: Float
dt = 0.10

time :: [Float]
time = iterate (+ dt) dt

controlInput :: [Control]
controlInput = map (\x -> V2 (10 * sin x) (0.01)) time

initialState :: StatePose
initialState = StatePose (V4 0 0.1 0.0349 0.5236) (identity !!* 0.01 :: Covariance)

rmat :: Covariance
rmat = identity

qmat :: M22 Float
qmat = V2 (V2 0.1 0)
          (V2 0 0.3)

createMeasNoise :: Int -> Float -> Measurement
createMeasNoise i q = (sqrt <$> qt) !* randVec
  where randVec = V2 (randMeasure !! (i * 2)) (randMeasure !! (i * 2 + 1))
        qt = qmat !!* q

updateStateX :: Control -> State -> State
updateStateX u x = x + V4 (dt * u ^._x * cos (x ^._w))
                          (dt * u ^._x * sin (x ^._w))
                          (dt * u ^._y)
                          (dt * u ^._x * tan (x ^._z + dt * u ^._y))

getStateTransition ::  Control -> State -> Covariance
getStateTransition u x = V4 (V4 1 0 0 (-dt * u ^._x * sin (x ^._w)))
                            (V4 0 1 0 (dt * u ^._x * cos (x ^._w)))
                            (V4 0 0 1 0 :: V4 Float)
                            (V4 0 0 (dt * u ^._x * ((tan (x ^._z + dt * u ^._y) ^ 2) + 1)) 1)


hFunc :: State -> Measurement
hFunc x = V2 (sqrt $ (x ^._x) ^ 2 + (x ^._y) ^ 2) (atan (x ^._y / x ^._x))

hJacobian :: State -> M24 Float
hJacobian xVec = V2 (V4 (x / sqrtxy) (y / sqrtxy) 0 0)
                    (V4 (-y / sqr_xy) (x / sqr_xy) 0 0)
  where x = xVec ^._x
        y = xVec ^._y
        sqr_xy = x ^ 2 + y ^ 2
        sqrtxy = sqrt sqr_xy

predict :: Control -> StatePose -> Float -> StatePose
predict u sp r = StatePose (updateStateX u (stateX sp)) ((f !*! p !*! (transpose f)) !+! rt)
  where p = stateP sp
        rt = rmat !!* r
        f = getStateTransition u (stateX sp)

update :: Measurement -> StatePose -> Float -> StatePose
update z sp q = StatePose (x + k !* y) (((identity :: M44 Float) !-! (k !*! h)) !*! p)
  where x = stateX sp           :: State
        p = stateP sp           :: Covariance
        y = z - (hFunc x)       :: Measurement
        h = hJacobian x         :: M24 Float
        b = (h !*! p !*! (transpose h)) !+! (qmat !!* q) :: M22 Float
        k = p !*! (transpose h) !*! (inv22 b) :: M42 Float

fakeMeasurements :: State -> Measurement -> Measurement
fakeMeasurements x noise = hFunc x + noise

kfStep :: Control -> Measurement -> StatePose -> Float -> Float -> StatePose
kfStep u mNoise sp r q = update z sp' q
  where sp' = predict u sp r
        z = fakeMeasurements (stateX sp') mNoise

toStateX :: StatePose -> (Float, Float)
toStateX sp = (stateX sp ^._x, stateX sp ^._y)

printState :: StatePose -> IO ()
printState sp = putStrLn ("X: " ++ show x ++ " Y: " ++ show y)
  where x = (stateX sp) ^._x
        y = (stateX sp) ^._y

repeatKFStep :: Int -> StatePose -> [Control] -> Float -> Float -> [StatePose]
repeatKFStep n' sp' inp' r q = repeatKFStep' n' sp' inp' r q 0

repeatKFStep' :: Int -> StatePose -> [Control] -> Float -> Float -> Int -> [StatePose]
repeatKFStep' n sp inp r q i
  | i == n = []
  | otherwise = sp' : repeatKFStep' n sp' fs r q (i+1)
  where (f : fs) = inp
        mNoise = createMeasNoise i q
        sp' = kfStep f mNoise sp r q
